File: backend/app/schemas/custom_runbook.py
"""
自定义 Runbook 请求/响应模型 (Custom Runbook Request/Response Models)
"""
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator


class RunbookStepSchema(BaseModel):
    """单个执行步骤"""
    name: str = Field(..., min_length=1, max_length=200)
    command: str = Field(..., min_length=1, max_length=2000)
    timeout_sec: int = Field(default=30, ge=1, le=3600)
    rollback_command: Optional[str] = Field(default=None, max_length=2000)


class CustomRunbookCreate(BaseModel):
    """创建自定义 Runbook"""
    name: str = Field(..., min_length=1, max_length=100)
    description: str = Field(default="", max_length=2000)
    trigger_keywords: List[str] = Field(default_factory=list)
    risk_level: str = Field(default="manual")
    steps: List[RunbookStepSchema] = Field(..., min_length=1)
    safety_checks: List[str] = Field(default_factory=list)
    is_active: bool = True

    @field_validator("risk_level")
    @classmethod
    def validate_risk_level(cls, v: str) -> str:
        allowed = {"auto", "confirm", "manual", "block"}
        if v not in allowed:
            raise ValueError(f"risk_level must be one of {allowed}")
        return v

    @field_validator("trigger_keywords")
    @classmethod
    def validate_keywords(cls, v: list) -> list:
        return [kw.strip() for kw in v if kw.strip()]


class CustomRunbookUpdate(BaseModel):
    """更新自定义 Runbook"""
    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    description: Optional[str] = Field(default=None, max_length=2000)
    trigger_keywords: Optional[List[str]] = None
    risk_level: Optional[str] = None
    steps: Optional[List[RunbookStepSchema]] = None
    safety_checks: Optional[List[str]] = None
    is_active: Optional[bool] = None

    @field_validator("risk_level")
    @classmethod
    def validate_risk_level(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            allowed = {"auto", "confirm", "manual", "block"}
            if v not in allowed:
                raise ValueError(f"risk_level must be one of {allowed}")
        return v


class CustomRunbookResponse(BaseModel):
    """自定义 Runbook 响应"""
    id: int
    name: str
    description: str
    trigger_keywords: List[str]
    risk_level: str
    steps: list
    safety_checks: List[str]
    created_by: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class RunbookListResponse(BaseModel):
    """Runbook 列表响应 (含内置+自定义)"""
    items: list
    total: int


class DryRunRequest(BaseModel):
    """Dry-run 请求"""
    variables: dict = Field(default_factory=dict)


class DryRunStepResult(BaseModel):
    """Dry-run 单步结果"""
    step_name: str
    resolved_command: str
    timeout_sec: int
    rollback_command: Optional[str] = None
    safety_check_passed: bool
    safety_message: str = "OK"


class DryRunResponse(BaseModel):
    """Dry-run 响应"""
    runbook_name: str
    risk_level: str
    total_steps: int
    steps: List[DryRunStepResult]
    all_safe: bool

File: frontend/src/services/customRunbook.ts
/**
 * 自定义 Runbook 服务模块
 * 提供自定义 Runbook 的 CRUD 和 Dry-run API 调用
 */
import api from './api';

/** Runbook 步骤 */
export interface RunbookStep {
  name: string;
  command: string;
  timeout_sec: number;
  rollback_command?: string | null;
}

/** 自定义 Runbook */
export interface CustomRunbook {
  id: number;
  name: string;
  description: string;
  match_alert_types: string[];
  trigger_keywords: string[];
  risk_level: string;
  steps: RunbookStep[];
  verify_steps: RunbookStep[];
  safety_checks: string[];
  created_by: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

/** Runbook 列表项 */
export interface RunbookListItem {
  id: number;
  name: string;
  description: string;
  source: 'custom';
  risk_level: string;
  match_alert_types?: string[];
  trigger_keywords?: string[];
  steps_count: number;
  is_active: boolean;
  created_at?: string;
}

/** Dry-run 步骤结果 */
export interface DryRunStepResult {
  step_name: string;
  resolved_command: string;
  timeout_sec: number;
  rollback_command?: string | null;
  safety_check_passed: boolean;
  safety_message: string;
}

export interface DryRunCheckResult {
  check: string;
  passed: boolean;
  message: string;
}

/** Dry-run 响应 */
export interface DryRunResponse {
  runbook_name: string;
  risk_level: string;
  total_steps: number;
  steps: DryRunStepResult[];
  preflight_checks: DryRunCheckResult[];
  all_safe: boolean;
}

/** 创建请求 */
export interface CreateRunbookRequest {
  name: string;
  description?: string;
  match_alert_types?: string[];
  trigger_keywords?: string[];
  risk_level?: string;
  steps: RunbookStep[];
  verify_steps?: RunbookStep[];
  safety_checks?: string[];
  is_active?: boolean;
}

/** 更新请求 */
export type UpdateRunbookRequest = Partial<CreateRunbookRequest>;

/** AI 生成 Runbook 请求 */
export interface GenerateRunbookRequest {
  description: string;
  risk_level?: string;
}

/** AI 生成 Runbook 响应 */
export interface GenerateRunbookResponse {
  success: boolean;
  runbook?: CreateRunbookRequest & { name: string; description: string };
  error?: string;
  safety_warnings: string[];
}

export const customRunbookService = {
  /** 获取所有 Runbook */
  listAll: () =>
    api.get<{ items: RunbookListItem[]; total: number }>('/runbooks/custom/all'),

  /** 获取自定义 Runbook 列表 */
  list: (params?: Record<string, unknown>) =>
    api.get<CustomRunbook[]>('/runbooks/custom', { params }),

  /** 获取单个自定义 Runbook */
  get: (id: number) => api.get<CustomRunbook>(`/runbooks/custom/${id}`),

  /** 创建自定义 Runbook */
  create: (data: CreateRunbookRequest) =>
    api.post<CustomRunbook>('/runbooks/custom', data),

  /** 更新自定义 Runbook */
  update: (id: number, data: UpdateRunbookRequest) =>
    api.put<CustomRunbook>(`/runbooks/custom/${id}`, data),

  /** 删除自定义 Runbook */
  delete: (id: number) => api.delete(`/runbooks/custom/${id}`),

  /** Dry-run */
  dryRun: (id: number, variables?: Record<string, string>) =>
    api.post<DryRunResponse>(`/runbooks/custom/${id}/dry-run`, { variables: variables || {} }),

  /** 导出 */
  exportAll: () =>
    api.get('/runbooks/custom/export/all', { responseType: 'blob' }),

  /** AI 生成 Runbook */
  generateWithAI: (data: GenerateRunbookRequest) =>
    api.post<GenerateRunbookResponse>('/ai/generate-runbook', data, { timeout: 60000 }),

  /** 导入 */
  importFile: (file: File) => {
    const formData = new FormData();
    formData.append('file', file);
    return api.post('/runbooks/custom/import', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
  },
};
